File: python/spectroxide/dark_photon.py
"""Helpers for dark-photon (γ ↔ A') conversion in the narrow-width approximation.

The PDE solver handles dark-photon oscillations through the initial-condition
path: pass
``injection={"type": "dark_photon_resonance", "epsilon": ε, "m_ev": m}`` to
:func:`spectroxide.solve` and the Rust solver computes ``γ_con``/``z_res``
itself, applying ``Δn(x) = −[1 − exp(−γ_con/x)] × n_pl(x)`` at ``z_start =
z_res`` and evolving forward in time. Use :func:`gamma_con` for standalone
diagnostics that need the conversion probability without running the PDE.

References
----------
- Mirizzi, Redondo & Sigl (2009), JCAP 0903, 026.
- Chluba, Cyr & Johnson (2024), MNRAS 535, 1874 [arXiv:2409.12115].
"""

from __future__ import annotations

from typing import Mapping, Tuple

import numpy as np
from scipy.optimize import brentq

from .greens import (
    _C_LIGHT,
    _HBAR,
    _K_BOLTZMANN,
    _cosmo_hubble,
    _cosmo_n_e,
)
from . import DEFAULT_COSMO, ionization_fraction

#: Type alias for cosmology mappings accepted by these helpers.
CosmoLike = Mapping[str, float]

_ALPHA_FS = 7.297_352_5693e-3
_M_ELECTRON = 9.109_383_7015e-31  # kg
_EV_IN_JOULES = 1.602_176_634e-19
_HBAR_EV_S = _HBAR / _EV_IN_JOULES


def plasma_frequency_ev(z: float, cosmo: CosmoLike | None = None) -> float:
    """Photon plasma frequency ``ω_pl`` at redshift ``z``.

    .. math::

        \\omega_{pl}^2 = \\frac{4 \\pi \\alpha\\, n_e\\, \\hbar c}{m_e},
        \\qquad n_e = X_e(z) \\, n_H(z).

    Parameters
    ----------
    z : float
        Redshift.
    cosmo : Mapping, optional
        Cosmological parameters.  Defaults to
        :data:`spectroxide.DEFAULT_COSMO`.

    Returns
    -------
    float
        Plasma frequency in **eV**.
    """
    if cosmo is None:
        cosmo = DEFAULT_COSMO
    n_e = _cosmo_n_e(z, cosmo)
    factor = 4.0 * np.pi * _ALPHA_FS * _HBAR * _C_LIGHT / _M_ELECTRON
    return _HBAR_EV_S * np.sqrt(n_e * factor)


def resonance_redshift(
    m_ev: float,
    cosmo: CosmoLike | None = None,
    z_min: float = 10.0,
    z_max: float = 3.0e7,
) -> float | None:
    """Solve ``ω_pl(z_res) = m`` for ``z_res`` by Brent bisection.

    Parameters
    ----------
    m_ev : float
        Dark-photon mass in **eV**.
    cosmo : Mapping, optional
        Cosmological parameters.  Defaults to
        :data:`spectroxide.DEFAULT_COSMO`.
    z_min : float, optional
        Lower edge of the search bracket (default ``10.0``).
    z_max : float, optional
        Upper edge of the search bracket (default ``3.0e7``).

    Returns
    -------
    float or None
        Resonance redshift, or *None* if no sign change of
        ``ω_pl(z) − m`` exists in ``[z_min, z_max]``.
    """
    if cosmo is None:
        cosmo = DEFAULT_COSMO

    def f(z):
        return plasma_frequency_ev(z, cosmo) - m_ev

    if f(z_min) * f(z_max) > 0:
        return None
    return brentq(f, z_min, z_max)


def dln_omega_pl_sq_dlna(z: float, cosmo: CosmoLike | None = None) -> float:
    """Compute ``|d ln ω_pl² / d ln a|`` at redshift ``z``.

    Uses a centered finite difference on ``X_e(z)`` with relative step
    ``max(0.1, 1e-4 z)``.  Falls back to the matter-era value 3 when
    ``X_e`` is too small to differentiate reliably.

    Parameters
    ----------
    z : float
        Redshift.
    cosmo : Mapping, optional
        Cosmological parameters.  Defaults to
        :data:`spectroxide.DEFAULT_COSMO`.

    Returns
    -------
    float
        Magnitude of the logarithmic derivative (dimensionless).
    """
    if cosmo is None:
        cosmo = DEFAULT_COSMO
    dz = max(0.1, z * 1e-4)
    x_e = ionization_fraction(z, cosmo)
    if x_e <= 1e-30:
        return 3.0
    x_e_p = ionization_fraction(z + dz, cosmo)
    x_e_m = ionization_fraction(z - dz, cosmo)
    dlnxe_dz = (x_e_p - x_e_m) / (2.0 * dz * x_e)
    return abs((1.0 + z) * dlnxe_dz + 3.0)


def gamma_con(
    epsilon: float, m_ev: float, cosmo: CosmoLike | None = None
) -> Tuple[float | None, float | None]:
    """Narrow-width approximation conversion parameter ``γ_con``.

    .. math::

        \\gamma_{con} = \\frac{\\pi\\, \\epsilon^2 m^2}
                            {\\bigl|d\\ln \\omega_{pl}^2 / d\\ln a\\bigr|_{z_{res}}
                             T_\\gamma(z_{res})\\, H(z_{res})},

    following Chluba & Cyr (2024) Eq. 6.

    Parameters
    ----------
    epsilon : float
        Kinetic-mixing parameter ``ε`` (dimensionless).
    m_ev : float
        Dark-photon mass in **eV**.
    cosmo : Mapping, optional
        Cosmological parameters.  Defaults to
        :data:`spectroxide.DEFAULT_COSMO`.

    Returns
    -------
    tuple of (float, float) or (None, None)
        ``(γ_con, z_res)`` if a resonance exists in the search bracket,
        otherwise ``(None, None)``.  ``γ_con`` is dimensionless;
        ``z_res`` is the resonance redshift.
    """
    if cosmo is None:
        cosmo = DEFAULT_COSMO
    z_res = resonance_redshift(m_ev, cosmo)
    if z_res is None:
        return None, None
    t_cmb_ev = _K_BOLTZMANN * cosmo["t_cmb"] * (1.0 + z_res) / _EV_IN_JOULES
    h_ev = _HBAR_EV_S * _cosmo_hubble(z_res, cosmo)
    d = dln_omega_pl_sq_dlna(z_res, cosmo)
    gc = np.pi * epsilon**2 * m_ev**2 / (d * t_cmb_ev * h_ev)
    return gc, z_res


def gc_per_epsilon_sq(
    m_ev: float, cosmo: CosmoLike | None = None
) -> Tuple[float | None, float | None]:
    """Return ``(γ_con/ε², z_res)`` at the resonance for mass ``m``.

    Convenience scaling factor: re-fit constraints on ``ε`` without
    recomputing ``z_res`` for each candidate.

    Parameters
    ----------
    m_ev : float
        Dark-photon mass in **eV**.
    cosmo : Mapping, optional
        Cosmological parameters.  Defaults to
        :data:`spectroxide.DEFAULT_COSMO`.

    Returns
    -------
    tuple of (float, float) or (None, None)
        ``(γ_con / ε² , z_res)``.  Returns ``(None, None)`` when no
        resonance exists.
    """
    return gamma_con(1.0, m_ev, cosmo)

File: src/dark_photon.rs
//! Helpers for dark photon (γ ↔ A') conversion in the narrow-width approximation.
//!
//! The canonical way to model γ ↔ A' resonant conversion in the PDE solver is
//! [`crate::energy_injection::InjectionScenario::DarkPhotonResonance`], which
//! takes (ε, m_{A'}) and internally calls [`gamma_con`] / [`resonance_redshift`]
//! to install the impulsive depletion
//! Δn(x) = -[1 - exp(-γ_con/x)] × n_pl(x) at `z_start = z_res`.
//!
//! The 1/x factor in that IC captures the frequency dependence of the
//! conversion probability P(x) ∝ 1/ω for ultrarelativistic photons.
//!
//! References:
//! - Mirizzi, Redondo & Sigl (2009), JCAP 0903, 026
//! - Chluba, Cyr & Johnson (2024), MNRAS 535, 1874
//! - Arsenadze et al. (2025), JHEP 03, 018

use crate::constants::*;
use crate::cosmology::Cosmology;
use crate::recombination::ionization_fraction;

/// Photon plasma frequency ω_pl (in eV) at redshift `z`.
///
/// ω_pl² = 4π α n_e ℏ c / m_e, with n_e = X_e(z) × n_H(z).
pub fn plasma_frequency_ev(z: f64, cosmo: &Cosmology) -> f64 {
    let hbar_ev_s = HBAR / EV_IN_JOULES;
    let x_e = ionization_fraction(z, cosmo);
    let n_e = cosmo.n_e(z, x_e);
    let factor = 4.0 * std::f64::consts::PI * ALPHA_FS * HBAR * C_LIGHT / M_ELECTRON;
    hbar_ev_s * (n_e * factor).sqrt()
}

/// Resonance redshift `z_res` where ω_pl(z_res) = m.
///
/// Returns `None` when `m` is outside the range spanned by ω_pl on
/// `[z_min, z_max] = [10, 3e7]`.
pub fn resonance_redshift(m_ev: f64, cosmo: &Cosmology) -> Option<f64> {
    let z_min = 10.0_f64;
    let z_max = 3.0e7_f64;
    let f = |z: f64| plasma_frequency_ev(z, cosmo) - m_ev;
    let (f_lo, f_hi) = (f(z_min), f(z_max));
    if !f_lo.is_finite() || !f_hi.is_finite() || f_lo * f_hi > 0.0 {
        return None;
    }
    let mut lo = z_min;
    let mut hi = z_max;
    let mut f_lo = f_lo;
    for _ in 0..80 {
        let mid = 0.5 * (lo + hi);
        let f_mid = f(mid);
        if (hi - lo) / mid < 1e-8 {
            return Some(mid);
        }
        if f_lo * f_mid <= 0.0 {
            hi = mid;
        } else {
            lo = mid;
            f_lo = f_mid;
        }
    }
    Some(0.5 * (lo + hi))
}

/// |d ln ω_pl² / d ln a| at redshift `z`. Centered finite difference in z.
pub fn dln_omega_pl_sq_dlna(z: f64, cosmo: &Cosmology) -> f64 {
    let dz = (z * 1.0e-4).max(0.1);
    let x_e = ionization_fraction(z, cosmo);
    if x_e <= 1e-30 {
        return 3.0;
    }
    let x_e_plus = ionization_fraction(z + dz, cosmo);
    let x_e_minus = ionization_fraction(z - dz, cosmo);
    let dlnxe_dz = (x_e_plus - x_e_minus) / (2.0 * dz * x_e);
    ((1.0 + z) * dlnxe_dz + 3.0).abs()
}

/// NWA dark-photon conversion parameter γ_con (dimensionless).
///
/// γ_con = π ε² m² / (|d ln ω_pl²/d ln a|_{z_res} × T_γ(z_res) × H(z_res)),
/// following Chluba & Cyr (2024) Eq. 6. Returns `None` if no resonance exists
/// in the supported redshift range.
///
/// Returned tuple: `(gamma_con, z_res)`.
pub fn gamma_con(epsilon: f64, m_ev: f64, cosmo: &Cosmology) -> Option<(f64, f64)> {
    let z_res = resonance_redshift(m_ev, cosmo)?;
    let t_cmb_ev = K_BOLTZMANN * cosmo.t_cmb * (1.0 + z_res) / EV_IN_JOULES;
    let hbar_ev_s = HBAR / EV_IN_JOULES;
    let h_ev = hbar_ev_s * cosmo.hubble(z_res);
    let d = dln_omega_pl_sq_dlna(z_res, cosmo);
    let gc = std::f64::consts::PI * epsilon * epsilon * m_ev * m_ev / (d * t_cmb_ev * h_ev);
    Some((gc, z_res))
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;

    #[test]
    fn plasma_frequency_matches_first_principles() {
        // Validates ω_pl = ℏ √(4πα n_e / m_e) against (a) an independent
        // first-principles computation using only CODATA constants + X_e, and
        // (b) the redshift scaling ω_pl ∝ (1+z)^{3/2} in the fully-ionized era.
        //
        // Reference: Mirizzi, Redondo & Sigl (2009), JCAP 0903, 026 Eq. 2.
        let cosmo = Cosmology::default();

        // Independent recomputation at z=1e5 (fully ionized, X_e≈1).
        let z = 1.0e5_f64;
        let x_e = ionization_fraction(z, &cosmo);
        assert!(x_e > 0.99, "z=1e5 should be fully ionized, got X_e={x_e}");
        let n_e = cosmo.n_e(z, x_e);
        let hbar_ev_s = HBAR / EV_IN_JOULES;
        let expected = hbar_ev_s
            * (n_e * 4.0 * std::f64::consts::PI * ALPHA_FS * HBAR * C_LIGHT / M_ELECTRON).sqrt();
        let omega = plasma_frequency_ev(z, &cosmo);
        let rel_err = (omega - expected).abs() / expected;
        assert!(
            rel_err < 1e-12,
            "ω_pl(1e5) = {omega:.4e} vs first-principles {expected:.4e}, rel_err={rel_err:.2e}"
        );

        // Verify (1+z)^{3/2} scaling in fully-ionized era where n_e ∝ (1+z)³:
        // ω_pl(z2) / ω_pl(z1) = [(1+z2)/(1+z1)]^{3/2}
        let z1 = 5.0e4_f64;
        let z2 = 2.0e5_f64;
        let ratio = plasma_frequency_ev(z2, &cosmo) / plasma_frequency_ev(z1, &cosmo);
        let expected_ratio = ((1.0 + z2) / (1.0 + z1)).powf(1.5);
        let scale_err = (ratio - expected_ratio).abs() / expected_ratio;
        assert!(
            scale_err < 1e-6,
            "ω_pl(1+z)^1.5 scaling violated: ratio={ratio:.6}, expected={expected_ratio:.6}"
        );
    }

    #[test]
    fn resonance_round_trip() {
        let cosmo = Cosmology::default();
        for m_ev in [3e-8, 1e-7, 1e-6, 1e-5, 1e-4] {
            let z_res = resonance_redshift(m_ev, &cosmo).expect("no resonance");
            let omega = plasma_frequency_ev(z_res, &cosmo);
            assert_relative_eq!(omega, m_ev, max_relative = 1e-5);
        }
    }

    #[test]
    fn gamma_con_scales_as_epsilon_squared() {
        let cosmo = Cosmology::default();
        let (g1, _) = gamma_con(1e-9, 1e-6, &cosmo).unwrap();
        let (g2, _) = gamma_con(2e-9, 1e-6, &cosmo).unwrap();
        assert_relative_eq!(g2 / g1, 4.0, max_relative = 1e-10);
    }

    #[test]
    fn gamma_con_matches_chluba_cyr() {
        // At m = 1e-7 eV the resonance sits in the fully-ionized era where
        // ω_pl ∝ (1+z)^{3/2}. With ω_pl(z=1e5) ≈ 5.50e-7 eV (see
        // plasma_frequency_matches_first_principles), setting ω_pl(z_res) = m:
        //   1+z_res = (1+1e5) × (1e-7 / 5.50e-7)^{2/3} ≈ 3.21e4
        //
        // For γ_con/ε², Chluba & Cyr (2024) Eq. 6 evaluated with this z_res,
        // default Planck cosmology, and X_e=1: expect ~2e10. Tolerance caps
        // parameter drift without being so tight that a switch from Planck
        // 2015→2018 ω_b would break the test.
        let cosmo = Cosmology::default();
        let (gc, z_res) = gamma_con(1.0, 1e-7, &cosmo).unwrap();
        assert!(
            (z_res - 3.21e4).abs() / 3.21e4 < 0.05,
            "z_res = {z_res:.3e}, expected ~3.21e4 (±5%) from ω_pl ∝ (1+z)^1.5"
        );
        // γ_con/ε² with default Planck cosmology and the resonance formula
        // gives 9.3e10 (measured); the tight 20% window catches a cosmology
        // parameter drift or prefactor bug, replacing the previous 100× window.
        assert!(
            (gc - 9.3e10).abs() / 9.3e10 < 0.2,
            "γ_con/ε² = {gc:.3e}, expected ~9.3e10 from Chluba & Cyr 2024 Eq. 6 (±20%)"
        );
    }
}
